srv: compilation error in openwrt buildsystem

I have a small C++ project that I want to extend with an ubus interface so that I can retrieve internal statistics via ubus.

So I gave ubuscpp a try and created an openwrt package so I can evaluate your c++ binding for ubus.
When compiling the examples (srv) I get the following error.

```
In file included from examples/srv.cpp:3:
examples/srv.cpp: In function 'int main(int, const char**)':
././include/ubus.hpp:28:2: error: no matching function for call to 'ubus_method::ubus_method(const char [5], main(int, const char**)::<lambda(ubus_context*, ubus_object*, ubus_request_data*, const char*, blob_attr*)>)'
   28 | })
      |  ^
examples/srv.cpp:22:25: note: in expansion of macro 'UBUS_HANDLER'
   22 |                         UBUS_HANDLER("test", ubus_test),
      |                         ^~~~~~~~~~~~
In file included from ././include/ubus.hpp:8:
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'ubus_method::ubus_method()'
  107 | struct ubus_method {
      |        ^~~~~~~~~~~
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 0 arguments, 2 provided
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'constexpr ubus_method::ubus_method(const ubus_method&)'
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 1 argument, 2 provided
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'constexpr ubus_method::ubus_method(ubus_method&&)'
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 1 argument, 2 provided
././include/ubus.hpp:28:2: error: no matching function for call to 'ubus_method::ubus_method(const char [4], main(int, const char**)::<lambda(ubus_context*, ubus_object*, ubus_request_data*, const char*, blob_attr*)>)'
   28 | })
      |  ^
examples/srv.cpp:23:25: note: in expansion of macro 'UBUS_HANDLER'
   23 |                         UBUS_HANDLER("get", ubus_get),
      |                         ^~~~~~~~~~~~
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'ubus_method::ubus_method()'
  107 | struct ubus_method {
      |        ^~~~~~~~~~~
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 0 arguments, 2 provided
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'constexpr ubus_method::ubus_method(const ubus_method&)'
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 1 argument, 2 provided
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'constexpr ubus_method::ubus_method(ubus_method&&)'
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 1 argument, 2 provided
././include/ubus.hpp:28:2: error: no matching function for call to 'ubus_method::ubus_method(const char [5], main(int, const char**)::<lambda(ubus_context*, ubus_object*, ubus_request_data*, const char*, blob_attr*)>)'
   28 | })
      |  ^
examples/srv.cpp:24:25: note: in expansion of macro 'UBUS_HANDLER'
   24 |                         UBUS_HANDLER("list", ubus_list),
      |                         ^~~~~~~~~~~~
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'ubus_method::ubus_method()'
  107 | struct ubus_method {
      |        ^~~~~~~~~~~
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 0 arguments, 2 provided
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'constexpr ubus_method::ubus_method(const ubus_method&)'
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 1 argument, 2 provided
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note: candidate: 'constexpr ubus_method::ubus_method(ubus_method&&)'
/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:107:8: note:   candidate expects 1 argument, 2 provided
examples/srv.cpp:21:34: error: cannot convert '<brace-enclosed initializer list>' to 'const std::vector<ubus_method>&'
   21 |                 srv -> add_object("ubus_test", {
      |                 ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
   22 |                         UBUS_HANDLER("test", ubus_test),
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   23 |                         UBUS_HANDLER("get", ubus_get),
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   24 |                         UBUS_HANDLER("list", ubus_list),
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   25 |                 });
      |                 ~~
././include/ubus.hpp:51:97: note:   initializing argument 2 of 'void ubus::service::add_object(std::string, const std::vector<ubus_method>&)'
   51 |                         void add_object(const std::string name, const std::vector<ubus_method> &methods);
      |                                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
make[3]: *** [Makefile:21: objs/srv.o] Error 1
make[3]: Leaving directory '/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/build_dir/target-mips_24kc_musl/ubuscpp-2023-09-30'
make[2]: *** [Makefile:52: /home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/build_dir/target-mips_24kc_musl/ubuscpp-2023-09-30/.built] Error 2
make[2]: Leaving directory '/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/feeds/tdt_thirdparty/libs/ubuscpp'
time: package/feeds/tdt_thirdparty/ubuscpp/compile#0.39#0.03#0.43
    ERROR: package/feeds/tdt_thirdparty/ubuscpp failed to build.
make[1]: *** [package/Makefile:128: package/feeds/tdt_thirdparty/ubuscpp/compile] Error 1
make[1]: Leaving directory '/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt'
make: *** [/home/feckert/workspace/openwrt/LDM-master-lantiq_xrx200/build/openwrt/include/toplevel.mk:232: package/feeds/tdt_thirdparty/ubuscpp/compile] Error 2
```

Can you give me a hint what the problem is?

Openwrt openwrt-21.02 with std=c++17 I get the following compilation build error.

```
mips-openwrt-linux-musl-g++ -Os -pipe -mno-branch-likely -mips32r2 -mtune=24kc -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -msoft-float -mips16 -minterlink-mips16 -fmacro-prefix-map=/home/feckert/workspace/openwrt/LDM-b_Otto/build/openwrt/build_dir/target-mips_24kc_musl/ubuscpp-2023-09-30=ubuscpp-2023-09-30 -Wformat -Werror=format-security -DPIC -fpic -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro  -I/home/feckert/workspace/openwrt/LDM-b_Otto/build/openwrt/staging_dir/toolchain-mips_24kc_gcc-8.4.0_musl/usr/include -I/home/feckert/workspace/openwrt/LDM-b_Otto/build/openwrt/staging_dir/toolchain-mips_24kc_gcc-8.4.0_musl/include/fortify -I/home/feckert/workspace/openwrt/LDM-b_Otto/build/openwrt/staging_dir/toolchain-mips_24kc_gcc-8.4.0_musl/include  --std=c++2a -Wall -fPIC -I./examples/include -I./jsoncpp/include -I././include -c -o objs/ubus.o src/ubus.cpp;
src/ubus.cpp: In member function 'void ubus::service::add_object(std::string, const std::vector<ubus_method>&)':
src/ubus.cpp:36:2: error: no match for 'operator=' (operand types are 'ubus_object' and '<brace-enclosed initializer list>')
  };
  ^
In file included from ././include/ubus.hpp:8,
                 from src/ubus.cpp:1:
/home/feckert/workspace/openwrt/LDM-b_Otto/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:125:8: note: candidate: 'constexpr ubus_object& ubus_object::operator=(const ubus_object&)'
 struct ubus_object {
        ^~~~~~~~~~~
/home/feckert/workspace/openwrt/LDM-b_Otto/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:125:8: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const ubus_object&'
/home/feckert/workspace/openwrt/LDM-b_Otto/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:125:8: note: candidate: 'constexpr ubus_object& ubus_object::operator=(ubus_object&&)'
/home/feckert/workspace/openwrt/LDM-b_Otto/build/openwrt/staging_dir/target-mips_24kc_musl/usr/include/libubus.h:125:8: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'ubus_object&&'
make[3]: *** [Makefile.inc:11: objs/ubus.o] Error 1
```

Looks like a casting error, hold on; as I don't have older compilers available, I can't test it, but I'll do something small to try fix it for you..

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 all: world
 
 CXX?=g++
-CXXFLAGS?=--std=c++23 -Wall -fPIC
+CXXFLAGS?=--std=c++17 -Wall -fPIC
 LDFLAGS?=-L/lib -L/usr/lib
 
 INCLUDES+= -I./examples/include -I./jsoncpp/include

--- a/examples/ubus_funcs.cpp
+++ b/examples/ubus_funcs.cpp
@@ -42,11 +42,16 @@ int ubus_test(const std::string& method, const std::string& msg, std::string& re
 	json::JSON args;
 
 	if ( !msg.empty()) {
+
 		std::error_code err;
 		args = json::JSON::Load(msg, err);
-		if ( args = json::JSON::Load(msg, err); !err )
+
+		if ( !err ) {
 			std::cout << "received args: " << args << std::endl;
-		else std::cout << "problem parsing " << msg << std::endl;
+		} else {
+			std::cout << "problem parsing " << msg << std::endl;
+		}
+
 	} else std::cout << "no arguments provided" << std::endl;
 
 	result = "{\"test\":" + args.dumpMinified() + "}";

--- a/include/ubus-cli.hpp
+++ b/include/ubus-cli.hpp
@@ -2,6 +2,7 @@
 
 #include <vector>
 #include <mutex>
+#include <bits/stdc++.h>
 
 extern "C" {
 #include <libubox/blobmsg_json.h>

--- a/include/ubus.hpp
+++ b/include/ubus.hpp
@@ -9,8 +9,8 @@ extern "C" {
 }
 
 #include "ubus-common.hpp"
-
-#define UBUS_HANDLER(name, func) ubus_method(name, [](ubus_context* ctx, ubus_object* obj, ubus_request_data* req, const char* method, blob_attr* msg) { \
+/*
+//#define UBUS_HANDLER(name, func) ubus_method(name, [](ubus_context* ctx, ubus_object* obj, ubus_request_data* req, const char* method, blob_attr* msg) { \
 	std::string _msg, _ret; \
 	\
 	if ( blob_len(msg) > 0 ) \
@@ -25,6 +25,29 @@ extern "C" {
 	} \
 	\
 	return code; \
+})
+*/
+
+#define UBUS_HANDLER(n, f) ubus_method({ \
+        .name = n, \
+        .handler = [](ubus_context* ctx, ubus_object* obj, ubus_request_data* req, const char* method, blob_attr* msg) { \
+                \
+                std::string _msg, _ret; \
+                \
+                if ( blob_len(msg) > 0 ) \
+                        _msg = std::string(blobmsg_format_json(msg, true)); \
+                \
+                int code = f(std::string(method), _msg, _ret); \
+                \
+                if ( _ret.size() != 0 ) { \
+                        blob_buf_init(&ubus::service::blob, 0); \
+                        blobmsg_add_json_from_string(&ubus::service::blob, _ret.c_str()); \
+                        ubus_send_reply(ctx, req, ubus::service::blob.head); \
+                } \
+                \
+                return code; \
+        }, \
+        .tags = 0 \
 })
 
 namespace ubus {

--- a/include/ubus.hpp
+++ b/include/ubus.hpp
@@ -9,24 +9,6 @@ extern "C" {
 }
 
 #include "ubus-common.hpp"
-/*
-//#define UBUS_HANDLER(name, func) ubus_method(name, [](ubus_context* ctx, ubus_object* obj, ubus_request_data* req, const char* method, blob_attr* msg) { \
-	std::string _msg, _ret; \
-	\
-	if ( blob_len(msg) > 0 ) \
-		_msg = std::string(blobmsg_format_json(msg, true)); \
-	\
-	int code = func(std::string(method), _msg, _ret); \
-	\
-	if ( _ret.size() != 0 ) { \
-		blob_buf_init(&ubus::service::blob, 0); \
-		blobmsg_add_json_from_string(&ubus::service::blob, _ret.c_str()); \
-		ubus_send_reply(ctx, req, ubus::service::blob.head); \
-	} \
-	\
-	return code; \
-})
-*/
 
 #define UBUS_HANDLER(n, f) ubus_method({ \
         .name = n, \

--- a/src/ubus-cli.cpp
+++ b/src/ubus-cli.cpp
@@ -27,7 +27,10 @@ static void make_receiver(ubus_context *ctx, int timeout) {
 
 	auto now = get_now(0);
 	receivers.remove_if([&now](const ubus_msg_receiver& entry) { return now > entry.expires; });
-	receivers.push_back(ubus_msg_receiver(ctx, get_now(timeout + 500)));
+	receivers.push_back(ubus_msg_receiver({
+		.ctx = ctx,
+		.expires = get_now(timeout + 500)
+	}));
 }
 
 static bool waiting_receiver_msg(ubus_context *ctx) {

--- a/src/ubus.cpp
+++ b/src/ubus.cpp
@@ -28,7 +28,7 @@ void ubus::service::add_object(const std::string name, const std::vector<ubus_me
 		.n_methods = (int)this -> methods.back().size()
 	};
 
-	this -> o = {
+	this -> o = (ubus_object){
 		.name = name.c_str(),
 		.type = &this -> t,
 		.methods = &this -> methods.back()[0],

--- a/src/ubus.cpp
+++ b/src/ubus.cpp
@@ -35,7 +35,9 @@ void ubus::service::add_object(const std::string name, const std::vector<ubus_me
 		.n_methods = (int)this -> methods.back().size()
 	};
 
-	if ( int ret = ubus_add_object(this -> ctx, &this -> o); ret != 0 )
+	int ret = ubus_add_object(this -> ctx, &this -> o);
+
+	if ( ret != 0 )
 		throw ubus::exception(ubus_strerror(ret), ret);
 	else this -> next_id++;
 

